tests/features/records-tags.spec.ts

```diff
@@ -0,0 +1,599 @@
+  describe('Records Tags', () => {
+    let didResolver: DidResolver;
+    let messageStore: MessageStore;
+    let dataStore: DataStore;
+    let eventLog: EventLog;
+    let eventStream: EventStream;
+    let dwn: Dwn;
+
+    // important to follow the `before` and `after` pattern to initialize and clean the stores in tests
+    // so that different test suites can reuse the same backend store for testing
+    before(async () => {
+      didResolver = new DidResolver({ didResolvers: [DidKey] });
+
+      const stores = TestStores.get();
+      messageStore = stores.messageStore;
+      dataStore = stores.dataStore;
+      eventLog = stores.eventLog;
+      eventStream = TestEventStream.get();
+
+      dwn = await Dwn.create({ didResolver, messageStore, dataStore, eventLog, eventStream });
+    });
+
+    beforeEach(async () => {
+      sinon.restore(); // wipe all previous stubs/spies/mocks/fakes
+
+      // clean up before each test rather than after so that a test does not depend on other tests to do the clean up
+      await messageStore.clear();
+      await dataStore.clear();
+      await eventLog.clear();
+    });
+
+    after(async () => {
+      await dwn.close();
+    });
+
+    describe('RecordsWrite with tags', () => {
+      it('should be able to write a Record tags', async () => {
+        const alice = await TestDataGenerator.generateDidKeyPersona();
+
+        // create tags that represent `string[]`, `number[]`, `string`, `number`, or `boolean` values.
+        const stringTag = 'string-value';
+        const stringArrayTag = [ 'string-value', 'string-value2' ];
+        const numberTag = 54566975;
+        const numberArrayTag = [ 0, 1 ,2 ];
+        const booleanTag = false;
+
+        const tagsRecord1 = await TestDataGenerator.generateRecordsWrite({
+          author    : alice,
+          published : true,
+          schema    : 'post',
+          tags      : {
+            stringTag,
+            numberTag,
+            booleanTag,
+            stringArrayTag,
+            numberArrayTag,
+          }
+        });
+
+        const tagsRecord1Reply = await dwn.processMessage(alice.did, tagsRecord1.message, { dataStream: tagsRecord1.dataStream });
+        expect(tagsRecord1Reply.status.code).to.equal(202);
+
+        // verify the record was written
+        const tagsRecord1Read = await RecordsRead.create({
+          filter: {
+            recordId: tagsRecord1.message.recordId,
+          },
+          signer: Jws.createSigner(alice)
+        });
+
+        const tagsRecord1ReadReply = await dwn.processMessage(alice.did, tagsRecord1Read.message);
+        expect(tagsRecord1ReadReply.status.code).to.equal(200);
+        expect(tagsRecord1ReadReply.record).to.not.be.undefined;
+        expect(tagsRecord1ReadReply.record!.descriptor.tags).to.deep.equal({ stringTag, numberTag, booleanTag, stringArrayTag, numberArrayTag });
+      });
+    });
+
+    describe('RecordsQuery filter for tags', async () => {
```

Comment:
Can we add a test or two on scenario of when tags are updated/deleted, the query would behavior correctly? (ie. query matching old/deleted values won't return anything and query matching new values will)

Reply:
Nah that's a good call out for a test scenario. I don't think it matters technically since the messageCid would be different, but who knows if some other indexing in the future would do something weird. Not bad to have those logical cases.

Reply:
Good call out! Although I tested updating and such on the index specific tests, I didn't have 'tags' in the mutable properties array, so the RecordsWrite as an update threw. Added it and some different tests on both the write, and the query side of things.

Reply:
Sanity: I see tests with queries after updates now, but don't see one for delete. I'll trust your judgement if you think delete is not necessary here, just sanity to make sure this is indeed intentional. Feel free to resolve this one as is.

Reply:
I ended up adding it now, it's not really necessary for query i don't think, but maybe good to have in case anything blows up with deleting a record w/ tags in general.